src/madonna/doxydocs/man/man3/madonna.3
.TH madonna.C 3 "28 Sep 2000" "madonna" \" -*- nroff -*-
.ad l
.nh
.SH NAME
madonna.C \- 
.SH SYNOPSIS
.br
.PP
\fC#include <iostream.h>\fR
.br
\fC#include <malloc.h>\fR
.br
\fC#include <stdlib.h>\fR
.br
\fC#include <libstruct.h>\fR
.br
\fC#include <sealibio.h>\fR
.br
\fC#include "genpart.h"\fR
.br
\fC#include "cost.h"\fR
.br
\fC#include "part.h"\fR
.br
\fC#include "globRoute.h"\fR
.br
\fC#include "phil_glob.h"\fR
.br
.SS Functions

.in +1c
.ti -1c
.RI "CIRCUITPTR \fBmadonna\fR (CIRCUIT *circuit)"
.br
.in -1c
.SS Variables

.in +1c
.ti -1c
.RI "int \fBrequestedGridPoints\fR [] = {NOTINITIALIZED, NOTINITIALIZED, NOTINITIALIZED}"
.br
.ti -1c
.RI "int \fBexpandableDirection\fR = NOTINITIALIZED"
.br
.ti -1c
.RI "LIBTABPTR \fBthislibtab\fR"
.br
.ti -1c
.RI "FUNTABPTR \fBthisfuntab\fR"
.br
.ti -1c
.RI "CIRTABPTR \fBthiscirtab\fR"
.br
.ti -1c
.RI "LAYTABPTR \fBthislaytab\fR"
.br
.ti -1c
.RI "LIBRARYPTR \fBthislib\fR"
.br
.ti -1c
.RI "FUNCTIONPTR \fBthisfun\fR"
.br
.ti -1c
.RI "CIRCUITPTR \fBthiscir\fR"
.br
.ti -1c
.RI "LAYOUTPTR \fBthislay\fR"
.br
.ti -1c
.RI "char* \fBcircuit_name\fR = NIL"
.br
.ti -1c
.RI "char * \fBfunction_name\fR = NIL"
.br
.ti -1c
.RI "char * \fBlibrary_name\fR = NIL"
.br
.ti -1c
.RI "char * \fBlayoutname\fR = NIL"
.br
.ti -1c
.RI "double \fBextraplaza\fR = 1.0"
.br
.ti -1c
.RI "int \fBmadonnamakeminiplaza\fR = TRUE"
.br
.ti -1c
.RI "int \fBmadonnamakepartition\fR = TRUE"
.br
.in -1c
.SH FUNCTION DOCUMENTATION
.PP 
.SS CIRCUITPTR madonna (CIRCUIT * circuit)
.PP
Definition at line 76 of file madonna.C.
.SH VARIABLE DOCUMENTATION
.PP 
.SS char * circuit_name = NIL
.PP
Definition at line 33 of file madonna.C.
.SS int expandableDirection = NOTINITIALIZED
.PP
Definition at line 20 of file madonna.C.
.SS double extraplaza = 1.0
.PP
Definition at line 37 of file madonna.C.
.SS char * function_name = NIL
.PP
Definition at line 34 of file madonna.C.
.SS char * layoutname = NIL
.PP
Definition at line 36 of file madonna.C.
.SS char * library_name = NIL
.PP
Definition at line 35 of file madonna.C.
.SS int madonnamakeminiplaza = TRUE
.PP
Definition at line 38 of file madonna.C.
.SS int madonnamakepartition = TRUE
.PP
Definition at line 39 of file madonna.C.
.SS int requestedGridPoints[] = {NOTINITIALIZED, NOTINITIALIZED, NOTINITIALIZED}
.PP
Definition at line 19 of file madonna.C.
.SS CIRCUITPTR thiscir
.PP
Definition at line 29 of file madonna.C.
.SS CIRTABPTR thiscirtab
.PP
Definition at line 24 of file madonna.C.
.SS FUNCTIONPTR thisfun
.PP
Definition at line 28 of file madonna.C.
.SS FUNTABPTR thisfuntab
.PP
Definition at line 23 of file madonna.C.
.SS LAYOUTPTR thislay
.PP
Definition at line 30 of file madonna.C.
.SS LAYTABPTR thislaytab
.PP
Definition at line 25 of file madonna.C.
.SS LIBRARYPTR thislib
.PP
Definition at line 27 of file madonna.C.
.SS LIBTABPTR thislibtab
.PP
Definition at line 22 of file madonna.C.
.SH AUTHOR
.PP 
Generated automatically by Doxygen for madonna from the source code.

src/RELEASENOTES

This file is maintained by: Arjan van Genderen
                            e-mail: arjan@cas.et.tudelft.nl

Sep. 22, 1999
-------------

Features compared to earlier distributions:
  
   - Seadali now also runs when it can allocate only a 
     reduced set of colors. (In the past, this caused
     that seadali could not be used at all on some
     screens.)
   - The maximum length of cell, instance, terminal 
     and net names has been increased from 14 to 256.
   - Net names are now used by trout to create labels
     in the layout.  As a result, the extracted circuit
     will have the same net (node) names as the original
     circuit.
   - Introduced an installation script 'ocean_install'.

Source code of this new OCEAN distribution is currently
not available.   The old OCEAN distribution is available 
from ftp://cas.et.tudelft.nl/pub/ocean.old.


Dec. 3, 1999
------------

Source code for compiling the cacd and OCEAN trees is now 
also available.  See the file COMPILATION.

Installation script was renamed oceanInstall.


Dec. 17, 1999
-------------

When the "no routing" option was used with trout, labels
were deleted from the layout.  This has now been fixed.


Jan. 5, 2000
------------

Another bug fix for handling labels during routing.


April 14, 2000
--------------

Fixed the problem that kissis didn't work when 
$SIS/sis_lib/.[sm]isrc are not available.


April 19, 2000
--------------

Fixed problem with directory names with more than 100 
characters.


June 8, 2000
------------

Fixed core dump for trout that appeared under Linux.


Sep. 12, 2000
-------------

- Added tools getldm and layflat.
- Added manual pages of layflat, fish, dofunc, mksls


Sep. 27, 2000
-------------

- Made extension in ICELLS for new cacd release that will use xcontrol 
  instead of macro and device


June 12, 2001
-------------

- Compiled with new version of cacd software.


March 13, 2002
--------------

- Compiled with new version of cacd software.
- Bug fix for seadali for deleting terminal in area.
- Let trout place and route master pins that are not connected to
  any instance.


August 12, 2002
---------------

- Updated celllibs/fishbone/src/nelsis3/template/sls_prototypes/oplib.ext



src/madonna/partitioner/madonna.C
//      @(#)madonna.C 1.13 06/30/94 Delft University of Technology
// 

#include <iostream.h>
#ifdef __MSDOS__
#include <alloc.h>
#endif
#include <malloc.h>
#include <stdlib.h>
#include <libstruct.h>
#include <sealibio.h>
#include "genpart.h"
#include "cost.h"
#include "part.h"
#include "globRoute.h"

#include "phil_glob.h"

extern int requestedGridPoints[];
extern int expandableDirection;

extern LIBTABPTR thislibtab;    /* set by existslib() */
extern FUNTABPTR thisfuntab;    /* set by existsfun() */
extern CIRTABPTR thiscirtab;    /* set by existscir() */
extern LAYTABPTR thislaytab;    /* set by existslay() */

extern LIBRARYPTR  thislib; /* set by sdfreadlib() sdfreadfun() sdfreadcir() sdfreadlay() */
extern FUNCTIONPTR thisfun; /* set by sdfreadfun() sdfreadcir() sdfreadlay() */
extern CIRCUITPTR  thiscir; /* set by sdfreadcir() sdfreadlay() */
extern LAYOUTPTR   thislay; /* set by sdfreadlay() */


extern char *circuit_name,
            *function_name,
            *library_name,
            *layoutname;
extern double extraplaza;
extern   int         madonnamakeminiplaza;
extern   int         madonnamakepartition;


static void printNetlistStatistics(CIRCUITPTR circuit)
{
   NETPTR net;
   int netdist[10],j,total_nets=0;
   for (j=0; j<10; ++j)
      netdist[j] = 0;
   for (net = circuit->netlist; net != NIL; net = net->next)
   {
      j = net->num_term;
      if (j < 0)
         ;
      else if (j <= 1)
         netdist[0] += 1;
      else if (j <= 7)
         netdist[j-1] += 1;
      else if (j <= 10)
         netdist[7] += 1;
      else if (j <= 20)
         netdist[8] += 1;
      else
         netdist[9] += 1;
   }
   for (j=0; j<10; ++j)
      total_nets += netdist[j];
   printf("------ net distribution (total #nets = %d):\n"
          "        0..1     2     3     4     5     6     7 8..10 11..20  >20\n\n"
          "      ",total_nets);
   for (j=0; j<10; ++j)
      printf(" %5d",netdist[j]);
   printf("\n\n");
}

/* Create a layout isomorph to the circuit. Call the layout "Madonna". */

CIRCUITPTR madonna(CIRCUIT *circuit)
{
   TOTALPPTR    total=NIL;
   FUNCTIONPTR  savethisfun=thisfun;
   CIRCUITPTR   savethiscir=thiscir,thecircuit,bestcircuit;
   LAYOUTPTR    savethislay=thislay;
   LIBRARYPTR   savethislib=thislib;
   
   initnetcostinfo();

   readImageFile();		    // Here we call the parser to read all 
                                    // the stuff about image
   
   printNetlistStatistics(circuit);

   if (madonnamakepartition)
   {
      madonna_(&total,circuit,1);

#ifdef __MSDOS__
      cerr << "\nAvailable memory: " << coreleft() << "\n\n";
#endif

      bestcircuit=(total ? total->bestpart : NIL);
   }
   else
      bestcircuit=NIL;

   if (total==NIL || total->bestpart==NIL)
      thecircuit=circuit;
   else
      thecircuit=total->bestpart;
   if (madonnamakeminiplaza )
   {

      if(! madonnamakepartition )
      {
	 phil(circuit_name,function_name,library_name,
	      layoutname,extraplaza,NULL,NULL,NULL);
      }
      else
      {
	 if (total != NIL && total->nx >= 1 && total->ny >= 1)
	 {
	    // ...and now for the global routing:
	    expansionGrid globgrid(total); // create global grid and globl nets
	    globgrid.routeGlobNets();	   // route the nets

	    phil(circuit_name,function_name,library_name,layoutname,
		 extraplaza,thecircuit,savethiscir,total->routing);
	 }
	 else
	 {
	    // nothing to do for the global router ...
	    phil(circuit_name,function_name,library_name,layoutname,
		 extraplaza,thecircuit,savethiscir,NIL);
	 }
      }
#ifdef __MSDOS__
      cerr << "\nAvailable memory: " << coreleft() << "\n\n";
#endif
   }
   thislay=savethislay; thiscir=savethiscir;
   thisfun=savethisfun; thislib=savethislib;
   
   return bestcircuit;
}
